pkg/kubelet/certificate/certificate_manager.go

```
@@ -0,0 +1,371 @@
+	return pairNamePrefix + "-" + qualifier + extension
+}
+
+// shouldRotate looks at how close the current certificate is to expiring and
+// decides if it is time to rotate or not.
+func (m *manager) shouldRotate() bool {
+	notAfter := m.cert.Leaf.NotAfter
+	total := notAfter.Sub(m.cert.Leaf.NotBefore)
+	remaining := notAfter.Sub(time.Now())
+	return (remaining * 100 / total) < shouldRotatePercent
+}
+
+func (m *manager) rotateCerts() error {
+	if !m.shouldRotate() {
+		return nil
+	}
+
+	ts := time.Now().Format("2006-01-02-15-04-05")
+	keyFilename := filename(ts, keyExtension)
+	certFilename := filename(ts, certExtension)
+
+	success := false
+
+	// Get the private key.
+	keyPath := filepath.Join(m.keyDirectory, keyFilename)
+	keyData, generatedKeyFile, err := certutil.LoadOrGenerateKeyFile(keyPath)
+	if err != nil {
+		return err
+	}
+	if generatedKeyFile {
+		defer func() {
+			if !success {
+				if err := os.Remove(keyPath); err != nil {
+					glog.Warningf("Cannot clean up the key file %q: %v", keyPath, err)
+				}
+			}
+		}()
+	}
+
+	// Call the Certificate Signing Request API to get a certificate for the
+	// new private key.
+	certPath := filepath.Join(m.certDirectory, certFilename)
+	certData, err := csr.RequestNodeCertificate(m.certSigningRequestClient, keyData, m.nodeName)
+	if err != nil {
+		return err
+	}
+	if err := certutil.WriteCert(certPath, certData); err != nil {
+		return err
+	}
+	defer func() {
+		if !success {
+			if err := os.Remove(certPath); err != nil {
+				glog.Warningf("Cannot clean up the cert file %q: %v", certPath, err)
+			}
+		}
+	}()
+
+	// Reload the cert/key data from disk so if there are any parsing or
+	// persisting problems they are discovered now before rotating to use the
+	// new certificate.
+	cert, err := loadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return err
+	}
+
+	designatedKeyPath := filepath.Join(m.keyDirectory, pairNamePrefix+keyExtension)
+	updateSymlink(keyPath, designatedKeyPath)
```

Don't discard the errors here (at least log them)